Add explicit TypeScript types throughout the app

obenland opened the pull request:
## Summary
- Add explicit TypeScript types for function parameters, return values, and variables
- Replace deprecated `JSX.Element` with `ReactNode` throughout the app
- Use type-only imports (`import type`) where appropriate
- Fix dependency comment structure (External/WordPress/Internal)
- Add jest-dom type declarations for test files

## Test plan
- [x] TypeScript compiles without errors (`npx tsc --noEmit`)
- [x] All unit tests pass (`npm run test:unit -- --testPathPattern="src/app"`)

obenland committed:
Add explicit TypeScript types throughout the app

- Replace deprecated JSX.Element with ReactNode across all components
- Add explicit type annotations for function parameters and returns
- Fix TypeScript errors with WordPress package type mismatches
- Fix test mock data to match updated type definitions
- Add jest-dom type declarations for test files
- Remove tsconfig paths mapping that caused source file type checking
- Use type assertions for TanStack Router and WordPress DataViews

obenland committed:
Merge remote-tracking branch 'origin/add/reader' into update/types-naming

# Conflicts:
#	src/app/components/actor-switcher/index.tsx
#	src/app/components/fields/avatar/avatar.tsx
#	src/app/components/fields/avatar/index.tsx
#	src/app/components/fields/metadata/__tests__/index.test.tsx
#	src/app/components/fields/metadata/index.tsx
#	src/app/routes/feed/__tests__/inspector.test.tsx
#	src/app/routes/feed/inspector.tsx

obenland committed:
Add explicit return types and fix type imports

- Add ReactNode return type to Avatar and Page components
- Use type-only imports for ReactNode in sidebar components
- Simplify onClick handler in SiteHub

obenland commented:
@pfefferle This is the cleanup PR we discussed. It includes [the `follower-actions` move](https://github.com/Automattic/wordpress-activitypub/pull/2624/changes#diff-dbae061db655ff7d5f5ae6e1cc97fcc551f28655ee635dfbd581897a3773db61). Please let me know if there's anything else that should be fixed as part of this one.

obenland committed:
Fix nullable type annotations

- getValue returns boolean | undefined when follow_status is optional
- find() returns FeedPost | undefined, not FeedPost
- getElementById returns HTMLElement | null

obenland commented on src/app/contexts/object-type-context.tsx:
```diff
@@ -8,26 +16,36 @@ interface ObjectTypeContextValue {
+}
+
+export function ObjectTypeProvider( { children }: ObjectTypeProviderProps ): ReactNode {
+	const { terms, isResolving } = useSelect( ( select ): { terms: Term[] | null; isResolving: boolean } => {
```
The explanation already exists in the comment above the CoreDataStoreSelectors interface definition on line 29-32. Adding it inline would be redundant.

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -42,20 +57,26 @@ const defaultLayouts = {
-		( id: number ) => {
-			navigate( {
-				search: ( prev: Record< string, unknown > ) => ( { ...prev, postId: id } ),
+	const selectItem: ( id: number ) => void = useCallback(
+		( id: number ): void => {
+			void navigate( {
+				search: ( ( prev: Record< string, unknown > ): Record< string, unknown > => ( {
+					...prev,
+					postId: id,
+				} ) ) as never,
```
This is a known limitation when using TanStack Router without strictNullChecks enabled globally. TanStack Router's type system requires strictNullChecks, which we can't enable project-wide without significant  refactoring. The as never cast is the documented workaround for this scenario.

obenland commented on src/app/routes/feed/inspector.tsx:
```diff
@@ -23,26 +28,34 @@ import { useTagFilter } from '../../hooks/use-tag-filter';
-				const { postId: _, ...rest } = prev as { postId?: number };
+	const onClose: () => void = (): void => {
+		void navigate( {
+			search: ( ( prev: Record< string, unknown > ): Record< string, unknown > => {
+				const { postId: _, ...rest } = prev as SearchParams;
 				return rest;
-			},
+			} ) as never,
```
This is a known limitation when using TanStack Router without strictNullChecks enabled globally. TanStack Router's type system requires strictNullChecks, which we can't enable project-wide without significant  refactoring. The as never cast is the documented workaround for this scenario.

obenland commented on src/app/store/resolvers.ts:
```diff
@@ -19,12 +20,12 @@ export function* getActiveActorId() {
 		const currentUser = select( coreStore ).getCurrentUser();
 		if ( currentUser?.id ) {
-			yield dispatch( STORE_NAME ).setActiveActor( currentUser.id );
+			yield ( dispatch( STORE_NAME ) as unknown as AppActions ).setActiveActor( currentUser.id );
```
WordPress data store types don't expose custom store actions from module augmentation within generator functions. The double cast is necessary because the standard dispatch(STORE_NAME) doesn't pick up our custom AppActions type in this context. This is a WordPress data typing limitation, not something we can fix without changes to @wordpress/data types.

obenland commented on src/app/routes/followers/stage.tsx:
```diff
@@ -65,7 +74,10 @@ export default function FollowersStage() {
+	const activeActorId: number = useSelect(
+		( select ): number => ( select( STORE_NAME ) as AppSelectors ).getActiveActorId(),
+		[]
+	);
```
The selector can technically return null, but this component only renders when an actor is active. The useFollowers hook handles the null case by disabling the query (enabled: false), so at the point this value is used, it's always a number.

obenland commented on src/app/tsconfig.json:
```diff
@@ -14,9 +14,6 @@
 		"resolveJsonModule": true,
 		"noEmit": true,
 		"baseUrl": ".",
-		"paths": {
-			"@wordpress/*": [ "../../node_modules/@wordpress/*/src" ]
-		},
 		"typeRoots": [ "../../node_modules/@types", "./types" ]
```
The paths mapping was added in the initial Social Web implementation (PR #2393) to resolve @wordpress/* imports to source files for type checking. However, this causes sprintf and other functions to use source types instead of the published type declarations, leading to type errors. Removing it allows TypeScript to use the proper type definitions from node_modules/@wordpress/*/build-types/. The build and type checking work correctly without it.

obenland commented on src/app/components/fields/content/index.tsx:
```diff
@@ -18,28 +30,28 @@ export const contentField: Field< FeedPost > = {
 		// Get the object type name from the cached map
-		const objectTypeId = item.ap_object_type?.[ 0 ];
-		const objectTypeName = getObjectTypeName( objectTypeId );
+		const objectTypeId: number = item.ap_object_type?.[ 0 ];
```
fixed in [c95f7e9](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/c95f7e9f83563bd2acb9163a2a291afff9f21420).

obenland committed:
Simplify type annotation.

obenland commented on src/app/components/sidebar/menu-description.tsx:
```diff
@@ -28,7 +28,7 @@ export default function MenuDescription( { menuItem: { description } }: MenuDesc
 	}
 
-	const Description = description;
+	const Description: ( () => ReactNode ) | ComponentType = description;
```
fixed in [f244b4d](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/f244b4d2dd7363885ffd846d52de2176287f4122).

obenland commented on src/app/index.tsx:
```diff
@@ -36,12 +40,12 @@ const routes: Route[] = [
  */
 export function initialize( id: string, settings: AppSettings ): void {
-	const target = document.getElementById( id );
+	const target: HTMLElement = document.getElementById( id );
```
Fixed in [c95f7e9](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/c95f7e9f83563bd2acb9163a2a291afff9f21420).

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -167,28 +188,30 @@ export default function FeedStage() {
+			const selectedId: string = nextSelection[ 0 ];
+			const selectedItem: FeedPost = feed.find(
+				( item: FeedPost ): boolean => item.id.toString() === selectedId
+			);
```
Fixed in [c95f7e9](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/c95f7e9f83563bd2acb9163a2a291afff9f21420)

obenland commented on src/app/components/fields/follow-status/index.tsx:
```diff
@@ -2,17 +2,29 @@
 	enableHiding: true,
-	getValue: ( { item }: { item: Actor } ) => item.follow_status?.follows_back,
-	render: ( { item }: { item: Actor } ) => {
+	getValue: ( { item }: { item: Actor } ): boolean => item.follow_status?.follows_back,
```
Fixed in [c95f7e9](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/c95f7e9f83563bd2acb9163a2a291afff9f21420).

obenland committed:
Fix more nullable type annotations

- objectTypeId can be undefined when ap_object_type array is empty
- objectTypeName returns string | null from getObjectTypeName
- activeActorId can be null from store's initial state

obenland commented on src/app/components/sidebar/feed-description.tsx:
```diff
@@ -22,9 +22,12 @@ import { STORE_NAME } from '../../store';
 export default function FeedDescription(): ReactNode {
-	const activeActorId = useSelect( ( select ) => ( select( STORE_NAME ) as AppSelectors ).getActiveActorId(), [] );
+	const activeActorId: number = useSelect(
+		( select ): number => ( select( STORE_NAME ) as AppSelectors ).getActiveActorId(),
```
Fixed in [04f0b4c](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/04f0b4c448d420739144aa37c0a37ce418a6911c).

obenland commented on src/app/components/fields/content/index.tsx:
```diff
@@ -18,28 +30,28 @@ export const contentField: Field< FeedPost > = {
-		const objectTypeId = item.ap_object_type?.[ 0 ];
-		const objectTypeName = getObjectTypeName( objectTypeId );
+		const objectTypeId: number = item.ap_object_type?.[ 0 ];
+		const objectTypeName: string = getObjectTypeName( objectTypeId );
```
Fixed in [04f0b4c](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/04f0b4c448d420739144aa37c0a37ce418a6911c).

obenland commented on src/app/components/fields/metadata/__tests__/index.test.tsx:
```diff
@@ -72,7 +86,7 @@ describe( 'metadataField', () => {
 			}
 
-			return render( <RenderComponent item={ post } />, { wrapper: Wrapper } );
+			return render( <RenderComponent item={ post } field={ metadataField as never } />, { wrapper: Wrapper } );
```
The as never cast in test files is needed because the test renders the field's render function directly rather than through DataViews. The Field type from @wordpress/dataviews has complex generic constraints that don't match when calling render in isolation. This is a test infrastructure limitation, not a production code issue.

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -42,20 +57,26 @@ const defaultLayouts = {
 	);
 	// Get active actor ID from store
-	const activeActorId = useSelect( ( select ) => ( select( STORE_NAME ) as AppSelectors ).getActiveActorId(), [] );
+	const activeActorId: number | null = useSelect(
+		( select ): number | null => ( select( STORE_NAME ) as AppSelectors ).getActiveActorId(),
+		[]
+	);
```
Fixed in [04f0b4c](https://github.com/Automattic/wordpress-activitypub/pull/2624/commits/04f0b4c448d420739144aa37c0a37ce418a6911c)

pfefferle commented:
What is the purpose of this PR? We seem that we handle JS very different than PHP!

obenland commented:
This is the cleanup PR we discussed in our 1:1. It includes [the follower-actions move](https://github.com/Automattic/wordpress-activitypub/pull/2624/changes#diff-dbae061db655ff7d5f5ae6e1cc97fcc551f28655ee635dfbd581897a3773db61).

> We seem that we handle JS very different than PHP!

I'd generally agree with that statement, is there anything specific that stands out to you?

pfefferle commented:
It feels like we’re very type-savvy in JavaScript now, but still quite vague in WordPress when it comes to typed function parameters and return values.

obenland commented:
I think that's mostly due to our back compat with PHP 7.2. Even upping that to 7.4 [could change that quite drastically](https://www.php.net/manual/en/language.types.declarations.php).

That said, you really want to have full control over the application if you heavily type your PHP code (which we don't have that as a WordPress plugin), as any type difference will cause a fatal error.

In summary: We handle it differently in JS because we can.

obenland committed:
Merge branch 'add/reader' into update/types-naming

obenland committed:
Fix TypeScript errors and improve type safety

- Add KeyboardEvent import to layout component
- Add react/react-dom types to tsconfig
- Remove yield from void function calls in resolvers
- Delete unused type files
